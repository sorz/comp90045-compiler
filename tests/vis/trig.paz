program trig;

{ Implementations of sin and cos using Taylor series expansion (E. Brumley) }

var testValue, sinOutput, cosOutput: real;

{ Calculate sin to approx 4 significant digits using Taylor series }
procedure sin(x: real; var out: real);
    var i, numIterations: integer;
        total, numerator, denominator, multiplier: real;

begin
    numIterations := 12;
    total := 0;
    i := 0;
    while (i < numIterations) do
    begin
        power(-1,i,numerator);
        factorial(2*i+1,denominator);
        power(x,(2*i+1),multiplier);
        total := total + (numerator / denominator) * multiplier;
        i := i + 1
    end;
    out := total
end;

{ Calculate cos to approx 4 significant digits using Taylor series }
procedure cos(x: real; var out: real);
    var i, numIterations: integer;
        total, numerator, denominator, multiplier: real;

begin
    numIterations := 12;
    total := 0;
    i := 0;
    while (i < numIterations) do
    begin
        power(-1,i,numerator);
        factorial(2*i,denominator);
        power(x,(2*i),multiplier);
        total := total + (numerator / denominator) * multiplier;
        i := i + 1
    end;
    out := total
end;
  
{ Calculate x to the power of n }
procedure power(x: real; n: integer; var out: real);
    var res: real;

begin
    if n = 0 then
        out := 1
    else
        if n = 1 then
            out := x
        else 
        begin
            power(x*x, n div 2, res);
            if 2*(n div 2) = n then 
                out := res        { n was even }
            else 
                out := x * res    { n was odd }
        end
end;

{ Calculate the factorial of x }
procedure factorial(x: integer; var out: real);

begin
    if x < 1 then
        out := 1
    else 
    begin
        out := x;
        while x > 1 do
        begin
            x := x - 1;
            out := out * x
        end
    end
end;

begin
    write('Test Value: ');
    read(testValue);
    sin(testValue, sinOutput);
    cos(testValue, cosOutput);
    writeln;

    write('Sin of ');
    write(testValue);
    write(' = ');
    write(sinOutput);
    writeln;

    write('Cos of ');
    write(testValue);
    write(' = ');
    write(cosOutput);
    writeln
end.

