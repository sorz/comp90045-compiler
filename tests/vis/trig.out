Test Value: 
Sin of 3.141593 = 0.000000
Cos of 3.141593 = -1.000000
